zingolib/src/wallet/transaction_records_by_id/trait_inputsource.rs

```diff
@@ -118,89 +122,70 @@ impl InputSource for TransactionRecordsById {
         anchor_height: zcash_primitives::consensus::BlockHeight,
         exclude: &[Self::NoteRef],
     ) -> Result<SpendableNotes<Self::NoteRef>, Self::Error> {
-        let mut sapling_note_noteref_pairs: Vec<(sapling_crypto::Note, NoteId)> = Vec::new();
-        let mut orchard_note_noteref_pairs: Vec<(orchard::Note, NoteId)> = Vec::new();
-        for transaction_record in self.values().filter(|transaction_record| {
-            transaction_record
-                .status
-                .is_confirmed_before_or_at(&anchor_height)
-        }) {
-            if sources.contains(&ShieldedProtocol::Sapling) {
-                sapling_note_noteref_pairs.extend(
-                    transaction_record
-                        .select_unspent_shnotes_and_ids::<SaplingDomain>()
-                        .into_iter()
-                        .filter(|note_ref_pair| !exclude.contains(&note_ref_pair.1)),
-                );
+        let mut unselected =
+            self.get_spendable_note_ids_and_values(sources, anchor_height, exclude);
+
+        unselected.sort_by_key(|(_id, value)| *value);
+
+        let mut selected = vec![];
+
+        let mut index_of_unselected = 0;
+        loop {
+            if unselected.is_empty() {
+                // all notes are selected. we pass the max value onwards whether we have reached target or not
+                break;
             }
-            if sources.contains(&ShieldedProtocol::Orchard) {
-                orchard_note_noteref_pairs.extend(
-                    transaction_record
-                        .select_unspent_shnotes_and_ids::<OrchardDomain>()
-                        .into_iter()
-                        .filter(|note_ref_pair| !exclude.contains(&note_ref_pair.1)),
-                );
+            match unselected.get(index_of_unselected) {
+                None => {
+                    // the iterator went off the end of the vector without finding a note big enough to complete the transaction... add the biggest note and reset the iteraton
+                    selected.push(unselected.pop().expect("nonempty"));
+                    index_of_unselected = 0;
+                    continue;
+                }
```

Comment:
I guess the argument for Some first is that we expect the first iteration of the loop to hit the Some case... however there are multiple ways to exit this loop
I normally put None first because it is a simpler case

Reply:
in this case, i typically expect the None case to notes before the if condition in the Some case to break up the loop. so i like this way